server/lib/buff_server/packages.ex

```diff
@@ -29,90 +30,98 @@ defmodule BuffServer.Packages do
   Gets a single package.
 
   Raises `Ecto.NoResultsError` if the Package does not exist.
-
-  ## Examples
-
-      iex> get_package!(123)
-      %Package{}
-
-      iex> get_package!(456)
-      ** (Ecto.NoResultsError)
-
   """
-  def get_package!(id) do
-    package = Repo.get!(Package, id)
-    package
-  end
+  def get_package!(id), do: Repo.get!(Package, id)
 
   @doc """
-  Creates a package.
-
-  ## Examples
-
-      iex> create_package(%{field: value})
-      {:ok, %Package{}}
-
-      iex> create_package(%{field: bad_value})
-      {:error, %Ecto.Changeset{}}
-
+  Creates the package and uploads to S3, this is run inside a transaction so if anything fails it'll rollback.
   """
-  def create_package!(%{artifact_binary: artifact_binary} = attrs) when is_binary(artifact_binary) do
-    changeset = %Package{}
-    |> Package.changeset(attrs, [:s3_bucket_name, :s3_bucket_path])
-
-    bucket_path = "/#{attrs.name}/artifact"
-    %{status_code: 200} = ExAws.S3.put_object(@s3_bucket_name, bucket_path, artifact_binary) |> ExAws.request!
+  def create_package(%{artifact_binary: artifact_binary} = attrs)
```

Comment:
Its usually more convenient to have method not throw

Reply:
MatchError occurs at compile time?

Reply:
hmm no that doesn't make any sense, however I don't see why the bang ! convention makes sense if any function case throw a MatchError.

Reply:
MatchError happens in run-time. 
yeah, you're kinda right. its super hard to really enforce this in a dynamic lang..
But the bang notation means the function itself shouldnt throw by default..(it still might, i guess :( )